drivers/flash/flash_simulator.c

```diff
@@ -51,7 +51,12 @@
 #error "Erase unit must be a multiple of program unit"
 #endif
 
+#ifndef CONFIG_FLASH_SIMULATOR_NOALLOC
 #define MOCK_FLASH(addr) (mock_flash + (addr) - FLASH_SIMULATOR_BASE_OFFSET)
+#else
+#define MOCK_FLASH(addr) (mock_flash + ((addr) % FLASH_SIMULATOR_PROG_UNIT) - \
+			  FLASH_SIMULATOR_BASE_OFFSET)
```

Comment:
This might go behind the address of `mock_flash`?

Reply:
Fixed.